Update of --config=mkl_aarch64 with Compute Library support

alenik01 opened the pull request:
This PR introduces the support of Compute Library for the Arm Architecture (ACL) as part of Bazel build for `mkl_aarch64` build config. The code was developed together with @joeramsay
The PR is related to
- previously raised [issue #47415](https://github.com/tensorflow/tensorflow/issues/47415) 
- PRs #47745 and #47679

Commit:
Update of --config=mkl_aarch64 with Compute Library support

This PR introduces the support of Compute Library for the Arm
Architecture (ACL) as part of Bazel build.

Signed-off-by: Aleksandr Nikolaev <aleksandr.nikolaev@arm.com>

alenik01 commented:
@agramesh1 @gzmkl please have a look.

gbaned commented:
@alenik01 Can you please resolve conflicts? Thanks!

alenik01 committed:
Merge branch 'master' into aarch64_build_patch

alenik01 commented:
@gbaned, GitHub GUI for the conflict resolution fails to pass google-cla bot for some reason. Should I force push?

penpornk commented:
@alenik01 Please don't force push. 

@google-cla I have verified that the commits are from the same user (@alenik01). Manually setting CLA to yes.

agramesh1 commented on tensorflow/core/kernels/mkl/mkl_conv_ops.cc:
```diff
@@ -657,6 +663,13 @@ class MklConvOp : public OpKernel {
+      convFwdDims.filter_address = const_cast<void*>(
+                                   static_cast<const void*>(
+                                   filter_tensor.flat<Tfilter>().data()));
+#endif
```
What is the rationale for adding the filter address to the key? If the filter is not a constant, the address could change from iteration to iteration.

alenik01 replied:
@agramesh1, this is due to Compute Library programming model, it creates a primitive per filter tensor, not per shape. And during the inference there is usually a set of constant addresses for the weights.

agramesh1 replied:
ok. then it should be fine.

agramesh1 commented on tensorflow/tensorflow.bzl:
```diff
@@ -365,6 +366,7 @@ def tf_copts(
         if_mkl(["-DINTEL_MKL=1"]) +
         if_mkldnn_openmp(["-DENABLE_ONEDNN_OPENMP"]) +
         if_enable_mkl(["-DENABLE_MKL"]) +
+        if_mkldnn_aarch64_acl(["-DENABLE_MKL", "-DENABLE_ONEDNN_OPENMP"]) +
```
This might be redundant, if_enable_mkl and if_mkldnn_openmp should be sufficient to set up the correct defines.

alenik01 replied:
Thanks for pointing this out, I will check and correct it.

alenik01 replied:
The line
`if_mkldnn_aarch64_acl(["-DENABLE_MKL", "-DENABLE_ONEDNN_OPENMP"])`
here is actually crucial for the correct functionality.

agramesh1 commented on third_party/mkl/build_defs.bzl:
```diff
@@ -77,6 +78,7 @@ def if_enable_mkl(if_true, if_false = []):
     """
     return select({
         "@org_tensorflow//third_party/mkl:enable_mkl": if_true,
+        "@org_tensorflow//third_party/mkl:build_with_mkl_aarch64": if_true,
```
This probably will result in a error if both the conditions are true. I don't think you need to check for build_with_mkl_aarch64 here since enable_mkl will be true.

alenik01 replied:
It indeed may be an overkill here, thanks.

cfRod replied:
Hi @agramesh1 , the reason for adding it here was that the enable_mkl() macro depends on enable=true and build_with_mkl=true and _I believe_ it is set to true if both conditions are set to true (https://github.com/tensorflow/tensorflow/blob/fe7c743b4734b29bc7fe12b3b09b4e8de4ecb5d4/third_party/mkl/BUILD#L56) . In our builds, we found that this that enable_mkl() is not activated under these conditions which means that even if we get a successful build with oneDNN it does pick oneDNN kernels in the verbose logs when running benchmarks.  The dependency on build_with_mkl in our case is false because we define a build_with_mkl_aarch64==true instead in .bazelrc . We can either re-define enable_mkl() here to activate with build_with_mkl_aarch64==true but that would again will work only if all three conditions are true which won't be the case. I suggest having a different enable_mkl_aarch64() instead. What do you think?

```
config_setting(
    name = "enable_mkl",
    define_values = {
        "enable_mkl": "true",
        "build_with_mkl": "true",
    },
    visibility = ["//visibility:public"],
)

```

In .bazelrc
```
build:mkl_aarch64 --define=build_with_mkl_aarch64=true --define=enable_mkl=true
build:mkl_aarch64 --define=tensorflow_mkldnn_contraction_kernel=0
build:mkl_aarch64 --define=build_with_mkl_opensource=true
```

alenik01 replied:
The line 
`"@org_tensorflow//third_party/mkl:build_with_mkl_aarch64": if_true,`
is indeed not needed here, we have this check earlier in this file in `if_mkl(if_true, if_false = [])`. I have tested, and everything actually builds fine without this check.

agramesh1 replied:
@cfRod and @alenik01 ok, you can leave it as it is for now.  As you can see we are restructuring the code now.
We will have to revisit this again soon.  There is a lot of legacy code leftover from a while ago and the names may no longer make sense, we plan to clean the build code shortly.

alenik01 replied:
@agramesh1, many thanks for the heads up, cleaning the build code will simplify the life significantly.

agramesh1 reviewed (changes requested):
Thanks @alenik01 for the PR. 
Also some of the files have missing newlines, can you fix those.

gbaned commented:
@alenik01  Can you please check @agramesh1's comments and keep us posted ? Thanks!

penpornk commented on third_party/compute_library/BUILD:
```diff
@@ -0,0 +1,79 @@
+    deps = ["arm_compute_core"],
+
+    visibility = ["//visibility:public"],
+)
```
Nit: Please add a new line at the end of file. Same for other files.

penpornk reviewed (changes requested):
Thank you for the PR! I have minor comments.

alenik01 commented on tensorflow/workspace2.bzl:
```diff
@@ -177,6 +177,28 @@ def _tf_repositories():
+       name = "mkl_dnn_acl_compatible",
+       build_file = "//third_party/mkl_dnn:mkldnn_acl.BUILD",
+       sha256 = "5f7fd92e2d0bf83580656695d4404e2cd1390ecad36496fd8ba10b5adc905f70",
+       strip_prefix = "oneDNN-2.1",
```
@penpornk, many thanks for the comments. I can also rebase this PR on the current TF master, what do you think about it?

penpornk replied:
@alenik01 Thank you for the quick reply!
If we upgrade `mkl_dnn_v1` to newer than v2.1, would that still be fine with your code?

penpornk replied:
Actually let's keep this separate repo (`mkl_dnn_acl_compatible`) for now, to be safe. Since the branch cut deadline is in 3 days. We should aim to get this PR in by today or tomorrow. (For any PR, there is a chance it breaks nightly tests and gets reverted the next day.)

alenik01 replied:
@penpornk,
>If we upgrade mkl_dnn_v1 to newer than v2.1, would that still be fine with your code?

yes, it will be fine. v2.2 is also a bit faster on AArch64.

penpornk reviewed (approved):
Since this PR doesn't seem to need any more significant changes, I'm going to approve this now and make the remaining minor fixes myself to save time. Thank you again for the PR!

alenik01 commented:
@penpornk, thanks. My main concern is how this PR stacks with recently merged PR [#47745](https://github.com/tensorflow/tensorflow/pull/47745). If the functionality will be broken, then what way would be the best to proceed, open another PR for the fix?

penpornk commented:
@alenik01 Since your build is a separate build, I don't think anything will clash (dependency-wise) as long as you keep `--define=tensorflow_mkldnn_contraction_kernel=0` in your build config. The reason I wanted to go with a separate repo was because `mkl_dnn_v1` v2.1 could still be reverted back to v1.6.4 at any time. (It's used by our official build so it goes through a large number of test configurations). So by having a separate repo, you can ensure you are able to use v2.1 regardless of what happens to `mkl_dnn_v1` (because we don't test `--config=mkl_aarch64`, so anything that passes your tests would be fine).

Now it's quite likely that `mkl_dnn_v1` would be able to keep v2.1. So if you'd like to go back to the same repo, that's fine too. Please let me know. I'll wait if you'd like to make more changes.

penpornk commented:
> If the functionality will be broken, then what way would be the best to proceed, open another PR for the fix?

Forgot to answer this one. Yes. If anything is broken after this PR, you can just send a PR to fix.

If your tests will be done by today, I think we can wait. Otherwise, I'd say we should get the PR in first.

alenik01 commented:
@penpornk, many thanks for the clarification.

> If your tests will be done by today

I will check and come back within 1.5 hours. My concern is about ENABLE_MKL macro, which is used by both PRs.

penpornk commented:
@alenik01 Sounds good. Thank you very much! :)

alenik01 committed:
Update tensorflow/workspace2.bzl

The URL with mirror.tensorflow.org for some reason produces ERROR 404: Not Found, the actual URL which is used during the build is https://github.com/ARM-software/ComputeLibrary/archive/v21.02.tar.gz

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

alenik01 committed:
Update BUILD

Newline added.

alenik01 committed:
Update mkldnn_acl.BUILD

alenik01 commented:
My minor fixes done within GitHub GUI are for some reason failing Google CLA checks despite I am in the corporate CLA group.

penpornk commented:
@googlebot I've verified that the other commits are from the same user, editing through Github's GUI. Manually changing CLA to yes.

penpornk reviewed (changes requested):
@alenik01 Please let me know when you are done with all the changes / this PR is ready to go. :) 

Please also resolve conflicts.

alenik01 committed:
Update build_defs.bzl

The second check of build_with_mkl_aarch64 indeed doesn't affect anything.

alenik01 commented:
@penpornk,
> Please also resolve conflicts.

I am resolving them right now.

penpornk reviewed (changes requested):
@alenik01 Oh, I mean resolving merge conflicts. The following files are currently out-of-date:
*  tensorflow/tensorflow.bzl
*  third_party/mkl/build_defs.bzl

alenik01 committed:
Merge branch 'master' into aarch64_build_patch

alenik01 commented:
@penpornk, the problem with TF is that it takes some time to build. Currently I am encountering the following error after the merge:
>create_tensorflow.python_api_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: dnnl_threadpool_interop_stream_create

Do you observe it for `--config=mkl` also?

penpornk reviewed (approved):
Thank you for the changes!

@googlebot I manually verified that all commits are from the same user. Reassigning the `cla: yes` label.

penpornk commented:
> @penpornk, the problem with TF is that it takes some time to build. Currently I am encountering the following error after the merge:
> 
> > create_tensorflow.python_api_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: dnnl_threadpool_interop_stream_create
> 
> Do you observe it for `--config=mkl` also?

@alenik01 I didn't see that with the default build. I don't think `--config=mkl` has seen that error, otherwise there CI test should have caught it. @agramesh1 Could you please help confirm? Thank you!

alenik01 commented:
@penpornk, this PR is ready to go in its current state, I have checked that with `--config=mkl_aarch64` Compute Library kernels are called and the correct accuracy is reproduced for ResNet50 inference. `mkl_dnn_acl_compatible` can later be combined with `mkl_dnn_v1` if you upgrade oneDNN version to 2.2 (we may probably need a separate PR for it).

@agramesh1, about `undefined symbol: dnnl_threadpool_interop_stream_create` it was false alarm, sorry for creating a mess.

penpornk commented:
@alenik01 Thank you for the update!

Could you please check the [Android Demo App](https://source.cloud.google.com/results/invocations/f06b044c-b836-4d2c-bee6-0d7d114da615) CI failure? It seems related:
```
tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc:22:10: fatal error: 'tensorflow/core/util/mkl_util.h' file not found
#include "tensorflow/core/util/mkl_util.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target //tensorflow/core/common_runtime/eager:execute failed to build
```

alenik01 committed:
Update build_defs.bzl

To test the effect on Android CI.

alenik01 committed:
Update build_defs.bzl

No "arm_any".

alenik01 commented:
Good morning, @penpornk. This Android CI failure seems to be out of scope of this PR, because according to the logs it is related to `android_arm64` build target, and also the header file `"tensorflow/core/util/mkl_util.h"` is in place. To proceed further the interaction with developers, who are responsible for `android_arm64` maintenance, is needed.

penpornk commented:
@alenik01 You're right. This break actually comes from #47745. Sorry about that! Fixing that now.

penpornk commented:
@alenik01 @agramesh1 I'm going to change `if_mkl` to just
```
    return select({
        "@org_tensorflow//third_party/mkl:build_with_mkl": if_true,
        "//tensorflow:linux_x86_64": if_true,
        "//tensorflow:windows": if_true,
        "//conditions:default": if_false,
    })
```
to avoid breaking builds on platforms TF-oneDNN isn't tested on. (It broke `//tensorflow:android_x86_64` and `//tensorflow:android_arm64`.)

@alenik01 I saw that you removed `"//tensorflow:arm_any": if_false,` from the list in this PR. Have you tested with all targets that  `//tensorflow:arm_any` covers? E.g.,
```
        ":arm",
        ":armeabi",
        ":armeabi-v7a",
        ":arm64-v8a",
        ":linux_aarch64",
        ":linux_armhf",
```

agramesh1 commented:
> @alenik01 @agramesh1 I'm going to change `if_mkl` to just
> 
> ```
>     return select({
>         "@org_tensorflow//third_party/mkl:build_with_mkl": if_true,
>         "//tensorflow:linux_x86_64": if_true,
>         "//tensorflow:windows": if_true,
>         "//conditions:default": if_false,
>     })
> ```
> 
> to avoid breaking builds on platforms TF-oneDNN isn't tested on. (It broke `//tensorflow:android_x86_64` and `//tensorflow:android_arm64`.)
> 
> @alenik01 I saw that you removed `"//tensorflow:arm_any": if_false,` from the list in this PR. Have you tested with all targets that `//tensorflow:arm_any` covers? E.g.,
> 
> ```
>         ":arm",
>         ":armeabi",
>         ":armeabi-v7a",
>         ":arm64-v8a",
>         ":linux_aarch64",
>         ":linux_armhf",
> ```

@penpornk Thanks that should work.  I think also adding back  "//tensorflow:arm_any": if_false, should also fix the android build issue.

gzmkl commented:
@penpornk please let me know if you need any help. Thanks!

penpornk commented:
> I think also adding back "//tensorflow:arm_any": if_false, should also fix the android build issue.

@agramesh1 `//tensorflow:arm_any` doesn't include `//tensorflow:android_arm64` and `//tensorflow:android_x86_64` which are breaking. `"//tensorflow:android": if_false` can fix the issue. But there are simply too many platforms to account for (`chromiumos`, `enscripten`, `linux_mips64`, etc) and there could be more in the future. It's too dangerous to match `//conditions:default` with `if_true`.

This PR is setting `//tensorflow:arm_any` to true. So I'm just asking to make sure that the `--config=aarch64` build has been tested on all the platforms `//tensorflow:arm_any` covers.
```
        "//tensorflow:arm",
        "//tensorflow:armeabi",
        "//tensorflow:armeabi-v7a",
        "//tensorflow:arm64-v8a",
        "//tensorflow:linux_aarch64",
        "//tensorflow:linux_armhf",
```
If not, it's probably better to break `//tensorflow:arm_any` down to only those that have been tested.

penpornk commented:
@gzmkl Thank you for the offer! The fix is being submitted. It should show up soon.

penpornk commented:
The fix is in (https://github.com/tensorflow/tensorflow/commit/37c1f29d1510c715ad1b6a249f32ac162b7fc2ec). 

@alenik01 Could you please help resolve merge conflicts again? Thank you very much!

alenik01 commented:
@penpornk, `arm_any` leads to the following conflict during the build:
>ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: /tensorflow/tools/pip_package/BUILD:179:10: Illegal ambiguous match on configurable attribute "data" in //tensorflow/tools/pip_package:licenses:
//third_party/mkl:build_with_mkl_aarch64
//tensorflow:arm_any

thus I removed it. May you please advise on the possible fix of "Illegal ambiguous match on configurable attribute"? The only new LICENSE file which this PR adds is in `./third_party/compute_library/`, and I don't see from where this "ambiguous match" arises.

>Have you tested with all targets that //tensorflow:arm_any covers?

The tests were performed with `linux_aarch64` and `arm64-v8a`.

alenik01 commented:
@penpornk, the new fix 37c1f29 actually excludes `arm_any` from if_mkl and mkl_deps.

>Could you please help resolve merge conflicts again?

I am now building TF to test.

alenik01 committed:
Merge branch 'master' into aarch64_build_patch

alenik01 commented:
>Could you please help resolve merge conflicts again?

Done.

penpornk reviewed (approved):
Thank you very much!

agramesh1 commented:
Hi @penpornk  sorry for the confusion. This commit https://github.com/tensorflow/tensorflow/commit/37c1f29d1510c715ad1b6a249f32ac162b7fc2ec broke the oneDNN builds due to ambiguous definitions of 
//third_party/mkl:build_with_mkl
//tensorflow:linux_x86_64

We verified that this change fixes it. Let me know if you can make the change or should we submit a PR. Thanks.

diff --git a/third_party/mkl/build_defs.bzl b/third_party/mkl/build_defs.bzl
index d913f22a110..7a4f08ed6a3 100644
--- a/third_party/mkl/build_defs.bzl
+++ b/third_party/mkl/build_defs.bzl
@@ -32,7 +32,6 @@ def if_mkl(if_true, if_false = []):
       may need it. It may be deleted in future with refactoring.
     """
     return select({
-        "@org_tensorflow//third_party/mkl:build_with_mkl": if_true,
         "//tensorflow:linux_x86_64": if_true,
         "//tensorflow:windows": if_true,
         "//conditions:default": if_false,
@@ -101,7 +100,6 @@ def mkl_deps():
       inclusion in the deps attribute of rules.
     """
     return select({
-        "@org_tensorflow//third_party/mkl:build_with_mkl": ["@mkl_dnn_v1//:mkl_dnn"],
         "@org_tensorflow//third_party/mkl:build_with_mkl_aarch64": ["@mkl_dnn_v1//:mkl_dnn_aarch64"],
         "//tensorflow:linux_x86_64": ["@mkl_dnn_v1//:mkl_dnn"],
         "//tensorflow:windows": ["@mkl_dnn_v1//:mkl_dnn"],

penpornk commented:
@agramesh1 Thank you for the patch! I'll make the change.

@googlebot I've verified that all the commits are from the same github user. Manually setting CLA to yes.

penpornk commented:
The fix is in. https://github.com/tensorflow/tensorflow/commit/5eabb16de7c41a115ccac4192c81a41487b676d0

No need to resolve merge conflicts again. I've already pulled this PR in and running tests internally.

cfRod commented:
Thanks everyone for sorting this out!
